azurelinuxagent/ga/policy/policy_engine.py

```diff
@@ -41,15 +70,157 @@ def _log_policy(cls, msg, is_success=True, op=WALAEventOperation.Policy, send_ev
+                        .format(type(allowlist_policy).__name__))
+                template["extensionPolicies"]["allowListedExtensionsOnly"] = allowlist_policy
+
+            # Validate and update template with global signature policy
+            signature_policy = extension_policies.get("signatureRequired")
+            if signature_policy is not None:
+                if not isinstance(signature_policy, bool):
+                    raise PolicyInvalidError(
+                        "invalid type {0} for attribute 'signatureRequired', please change to bool"
+                        .format(type(signature_policy).__name__))
+                template["extensionPolicies"]["signatureRequired"] = signature_policy
+
+            # Parse individual extension policies
+            extensions = extension_policies.get("extensions")
+            if extensions is not None:
+                for ext_name, ext_policy in extensions.items():
+                    if ext_policy is not None:
+                        individual_signing_policy = ext_policy.get("signatureRequired")
+                        if individual_signing_policy is not None:
+                            if not isinstance(individual_signing_policy, bool):
+                                raise PolicyInvalidError(
+                                    "invalid type {0} for attribute 'signatureRequired', please change to bool."
+                                    .format(type(individual_signing_policy).__name__))
+                        else:
+                            # If individual extension is present but signature policy not specified, use global policy.
+                            individual_signing_policy = template.get("extensionPolicies").get("signatureRequired")
+
+                        # Build individual extension policy and add to template.
+                        policy_to_add = {
+                            "signatureRequired": individual_signing_policy
+                        }
+                        template["extensionPolicies"]["extensions"][ext_name] = policy_to_add
+
+                # Convert "extensions" to a case-folded dict for case-insensitive lookup
+                case_folded_extension_dict = _CaseFoldedDict.from_dict(template["extensionPolicies"]["extensions"])
+                template["extensionPolicies"]["extensions"] = case_folded_extension_dict
+
+        return template
+
+
+class ExtensionPolicyEngine(_PolicyEngine):
+
+    def should_allow_extension(self, extension_to_check):
+        """
+        Return whether we should allow extension download based on policy.
+        extension_to_check is expected to be an Extension object.
+
+        If policy feature not enabled, return True.
+        If allowListedExtensionsOnly=true, return true only if extension present in "extensions" allowlist.
+        If allowListedExtensions=false, return true always.
+        """
+        if not self.is_policy_enforcement_enabled():
```

Comment:
is_policy_enforement_enabled() result could change between the processing of each extension in a single goal state, if customer deletes the customer policy file. 

Should we check this state once in init and keep it as an instance variable?

Reply:
Good point. PolicyEngine will be instantiated on a per goal state basis, so makes sense to check right before reading the policy file and store it as an instance variable. I've made this change